lib/rucio/core/did_meta_plugins/elasticsearch_meta.py

```diff
@@ -0,0 +1,406 @@
+# Copyright European Organization for Nuclear Research (CERN) since 2012
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Author(s):
+# - Anil Panta <anilpanta2@gmail.con>, 2023
+
+'''
+ Elasticsearch based metadata plugin.
+'''
+
+import datetime
+import operator
+from typing import TYPE_CHECKING, Any, Optional, Union
+
+from elasticsearch import Elasticsearch
+from elasticsearch import exceptions as elastic_exceptions
+
+from rucio.common import config, exception
+from rucio.core.did_meta_plugins.did_meta_plugin_interface import DidMetaPlugin
+from rucio.core.did_meta_plugins.filter_engine import FilterEngine
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from sqlalchemy.orm import Session
+
+    from rucio.common.types import InternalScope
+
+IMMUTABLE_KEYS = [
+    'scope',            # generated on insert
+    'name',             # generated on insert
+    'vo'                # generated on insert
+]
+
+
+class ElasticDidMeta(DidMetaPlugin):
+    def __init__(
+        self,
+        hosts: "Optional[list[str]]" = None,
+        user: "Optional[str]" = None,
+        password: "Optional[str]" = None,
+        index: "Optional[str]" = None,
+        archive_index: "Optional[str]" = None,
+        use_ssl: "Optional[bool]" = False,
+        verify_certs: bool = True,
+        ca_certs: "Optional[str]" = None,
+        client_cert: "Optional[str]" = None,
+        client_key: "Optional[str]" = None,
+        request_timeout: int = 100,
+        max_retries: int = 3,
+        retry_on_timeout: bool = False
+    ) -> None:
+        super(ElasticDidMeta, self).__init__()
+        hosts = hosts or [config.config_get('metadata', 'elastic_service_host')]
```

Comment:
Should we only allow a single `elastic_service_host`, or should we change `metadata.elastic_service_hosts` to be a list where multiple hosts can be configured?